EIPS/eip-4399.md

```diff
@@ -0,0 +1,110 @@
+
+Additionally, this ensures that the execution layer can be fully executed with the block alone rather than requiring extra inputs from the PoS consensus layer.
+
+Mixing the randomness into a block header may contribute to uniqueness of the block hash in the case when values of other fields of the block header match the corresponding values of the header of another block.
+
+### Using `mixHash` field instead of `difficulty`
+
+Client software heavily depends on the `difficulty` value as total difficulty computation is the basis of the PoW fork choice rule. Setting the `difficulty` field to `0` at the PoS upgrade aims to reduce the surface of bugs related to the total difficulty value growing after the upgrade.
+
+Additionally, total difficulty computation would become overflow prone if the randomness output supplanted the value of the `difficulty` field.
+
+### Reusing existing field instead of appending a new one
+
+The `mixHash` field is deprecated at the PoS upgrade and set to zero bytes array thereafter. Reusing an existing field as a place for the randomness output saves 32 bytes per block and effectively removes the deprecation of one of the fields induced by the upgrade.
+
+### Reusing the `DIFFICULTY` opcode instead of introducing a new one
+
+See the [Motivation](#motivation).
+
+### Renaming the field and the opcode
+
+The renaming should be done to make the field and the opcode names semantically sound.
+
+
+## Backwards Compatibility
+
+This EIP doesn't introduce any backward incompatible changes to the base protocol. Though, the changes it propose might be backward incompatible for the following categories of applications:
+* Applications that uses the value returned by the `DIFFICULTY` opcode as the PoW `difficulty` parameter
+* Applications which logic depending on the `DIFFICULTY` opcode return value to be relatively small number
+
+The first category is already affected by switching the consensus mechanism to PoS and no additional breaking changes are introduced by this specification.
+
+The second category comprises the applications that use the return value of the `DIFFICULTY` opcode in operations that might cause either overflow or underflow errors. Applications where the result of such errors may turn into security violations are not considered by this EIP.
+
+
+## Test Cases
+
+Testing this functionality goes beyond this document and should be done in a form of integration testing.
+
+
+## Reference Implementation
+
+*TBD*
+
+
+## Security Considerations
+
+The `RANDOM` opcode in the PoS network should be considered as a source of randomness output of a higher strength than the output provided by either the `BLOCKHASH` or the `DIFFICULTY` opcode in the PoW network.
+
+However, the randomness output returned by the `RANDOM` opcode is revealed within the parent block.
```

Comment:
yes

and an early randao reveal slashing is on the table
was in fact in the phase 0 specs early on!

but it's a relatively low priority attack point so I don't expect we see a proposal for inclusion in the next 18 months, tbh

Reply:
My try to be more clear on it is: "is revealed to the network within the parent block" -- https://github.com/ethereum/EIPs/pull/4407. I slightly prefer us do not describe an attack scenario with the collusion of multiple proposers in this document. The reveal to the network should create enough level of awareness for smart contract developers. But I don't have a strong opinion on that and we could and it might worth specifying from some perspective.

Early randao reveal slashing doesn't protect revealing the randao mix, does it? If it does then revealing after grandparent block is publish can't be punished unless the proposer punishes itself.